File: backend/app/schemas/deploy.py
"""Pydantic schemas for the one-click deployment endpoints (admin)."""

from __future__ import annotations

import uuid as uuid_lib
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DeploymentTokenCreateRequest(BaseModel):
    label: str | None = Field(default=None, max_length=255)
    # None → never expires. Bounded to a year to avoid forgotten eternal tokens.
    expires_in_days: int | None = Field(default=None, ge=1, le=365)
    # None → unlimited machines may enrol with this one link.
    max_uses: int | None = Field(default=None, ge=1, le=100_000)


class DeploymentTokenCreated(BaseModel):
    uuid: uuid_lib.UUID
    token: str = Field(..., description="plaintext — show once, used to build the install link")
    label: str | None
    expires_at: datetime
    max_uses: int | None


class DeploymentTokenOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: uuid_lib.UUID
    label: str | None
    expires_at: datetime
    max_uses: int | None
    use_count: int
    revoked_at: datetime | None
    created_at: datetime


class BinaryInfoOut(BaseModel):
    """The agent binary currently being served — shown as a version badge on the
    deploy page so an admin can confirm which build a downloaded installer carries."""

    version: str
    os: str
    arch: str
    # sha256 ของ binary ที่เสิร์ฟอยู่ — เปลี่ยนทุกครั้งที่ build ใหม่ (ต่างจาก version
    # ที่ถูก fix เป็น 0.1.0 เสมอ) ใช้เป็น "build fingerprint" ให้ admin ยืนยันบนหน้า
    # deploy ได้ว่าตัวที่กำลังเสิร์ฟ/กำลังจะโหลด เป็นบิลด์ใหม่จริง ไม่ใช่ของเก่าค้าง
    sha256: str
    # build_stamp = UTC timestamp ของตอน build ซึ่งถูก "bake เข้า binary" ด้วย → installer
    # wizard โชว์ค่าเดียวกันนี้บนหน้าแรก ผู้ใช้จึงเทียบ "ค่าบนหน้า deploy" กับ "ค่าบนตัวที่
    # โหลดไปรัน" ได้ตรงๆ ตรงกัน = ตัวที่โหลด = build ล่าสุดแน่นอน ("" = manifest เก่า)
    build_stamp: str = ""

File: dashboard/lib/deploy.ts
"use client";

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api, BASE_URL } from "./api";

// ── Types (mirror backend app/schemas/deploy.py) ──────────────────────────────

export interface DeploymentTokenCreateInput {
  label?: string | null;
  // null → never expires (backend caps at 365 days when set)
  expires_in_days?: number | null;
  // null → unlimited machines may enrol with this one link
  max_uses?: number | null;
}

export interface DeploymentTokenCreated {
  uuid: string;
  token: string; // plaintext — shown once, used to build the install link
  label: string | null;
  expires_at: string;
  max_uses: number | null;
}

export interface DeploymentToken {
  uuid: string;
  label: string | null;
  expires_at: string;
  max_uses: number | null;
  use_count: number;
  revoked_at: string | null;
  created_at: string;
}

// ── Hooks ─────────────────────────────────────────────────────────────────────

export function useDeploymentTokens(includeRevoked = false) {
  return useQuery({
    queryKey: ["deployment-tokens", includeRevoked],
    queryFn: async () => {
      const { data } = await api.get<DeploymentToken[]>("/deploy/tokens", {
        params: { include_revoked: includeRevoked },
      });
      return data;
    },
  });
}

export function useCreateDeploymentToken() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (input: DeploymentTokenCreateInput): Promise<DeploymentTokenCreated> => {
      const { data } = await api.post<DeploymentTokenCreated>("/deploy/tokens", input);
      return data;
    },
    onSuccess: () => void qc.invalidateQueries({ queryKey: ["deployment-tokens"] }),
  });
}

export function useRevokeDeploymentToken() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (uuid: string) => {
      const { data } = await api.post<DeploymentToken>(`/deploy/tokens/${uuid}/revoke`);
      return data;
    },
    onSuccess: () => void qc.invalidateQueries({ queryKey: ["deployment-tokens"] }),
  });
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/** Default callback URL the agent will phone home to: the API base minus the
 *  `/api/v1` suffix (e.g. http://localhost:8001). Admins can override this with
 *  a LAN IP so enrolled machines can actually reach the backend. */
export function defaultServerUrl(): string {
  const base = BASE_URL.replace(/\/api\/v\d+\/?$/, "");
  if (typeof window !== "undefined" && base.startsWith("http://localhost")) {
    // localhost only resolves on the admin's own box — prefill with the host the
    // admin opened the dashboard from so the link is shareable by default.
    return `${window.location.protocol}//${window.location.hostname}:${new URL(base).port || "8001"}`;
  }
  return base;
}

/** Public installer download link. The token in the query authorizes it, so this
 *  can be opened directly by an end user's browser (no admin auth needed).
 *
 *  The download is served from the same `server` address the admin entered (the
 *  one machines must reach), NOT from the dashboard's own API base — otherwise a
 *  link opened on another machine would resolve `localhost` to itself and fail.
 */
export function installerUrl(
  token: string,
  server: string,
  os = "windows",
  arch = "amd64",
): string {
  const root = server.replace(/\/$/, "");
  const params = new URLSearchParams({ token, os, arch, server: root });
  return `${root}/api/v1/deploy/installer?${params.toString()}`;
}

// ── Status (pure — unit-tested in deploy.test.ts) ─────────────────────────────

export type DeployStatus = "active" | "revoked" | "expired" | "exhausted";

/** Backend stores year-2999 as the "never expires" sentinel. */
export function isNeverExpiry(expiresAt: string): boolean {
  return new Date(expiresAt).getUTCFullYear() >= 2999;
}

/** Effective state of a deployment link for badge display. */
export function deployStatus(token: DeploymentToken, now: Date = new Date()): DeployStatus {
  if (token.revoked_at) return "revoked";
  if (!isNeverExpiry(token.expires_at) && new Date(token.expires_at) <= now) {
    return "expired";
  }
  if (token.max_uses != null && token.use_count >= token.max_uses) return "exhausted";
  return "active";
}
